File: swift/InkRecognition/Recognizer/Recognizer/InkRecognitionUnit.swift

import Foundation
import UIKit

@objc
public enum InkRecognitionUnitCategory: Int {
    case unknown,
    inkWord,
    inkDrawing,
    inkBullet,
    listItem,
    paragraph,
    line,
    writingRegion
}

@objc
class InkBoundingRectangle : NSObject, Decodable {
    var topX: CGFloat = 0
    var topY: CGFloat = 0
    var width: CGFloat = 0
    var height: CGFloat = 0
    
    init(x: Float, y: Float, width: Float, height: Float) {
        self.topX = InkPoint.millimeterToCGFloat(mmValue: x)
        self.topY = InkPoint.millimeterToCGFloat(mmValue: y)
        self.width = InkPoint.millimeterToCGFloat(mmValue: width)
        self.height = InkPoint.millimeterToCGFloat(mmValue: height)
    }
}

@objc
class InkRecognitionUnit: NSObject {
    private var categoryString: String
    private var boundingRectangle: InkBoundingRectangle!
    private var rotatedBoundingRectangle = [CGPoint]()
    private var childIds = [Int]();
    private var parentId = -1
    private var strokeIds: [Int]
    public var id: Int = 0
    private var dotsPerInch: Float = 0.0
    private var result: InkRoot!
    
    @objc
    public var category: InkRecognitionUnitCategory {
        get {
            var recognitionCategory: InkRecognitionUnitCategory
            switch (self.categoryString) {
            case "inkWord":
                recognitionCategory = InkRecognitionUnitCategory.inkWord
            case "inkDrawing":
                recognitionCategory = InkRecognitionUnitCategory.inkDrawing
            case "inkBullet":
                recognitionCategory = InkRecognitionUnitCategory.inkBullet
            case "line":
                recognitionCategory = InkRecognitionUnitCategory.line
            case "listItem":
                recognitionCategory = InkRecognitionUnitCategory.listItem
            case "paragraph":
                recognitionCategory = InkRecognitionUnitCategory.paragraph
            case "writingRegion":
                recognitionCategory = InkRecognitionUnitCategory.writingRegion
            default:
                recognitionCategory = InkRecognitionUnitCategory.unknown
            }
            return recognitionCategory;
        }
    }
    
    @objc
    public var children: [InkRecognitionUnit] {
        return result.getNodes(ids: childIds)
    }
    
    @objc
    public var parent: InkRecognitionUnit {
        return result.getNode(id: parentId)
    }
    
    @objc
    public var boundingBox: InkBoundingRectangle {
        return self.boundingRectangle
    }
    
    @objc
    public var rotatedBoundingBox: [CGPoint] {
        return self.rotatedBoundingRectangle
    }
    
    @objc
    init(json: [String: Any]) {
        self.id = json["id"] as! Int
        self.parentId = json["parentId"] as! Int
        
        if let idsForChildren = json["childIds"] as? [Int] {
            self.childIds = idsForChildren
        }
        self.categoryString = json["category"] as! String
        self.strokeIds = json["strokeIds"] as! [Int]
        
        let jsonBoundingRect = json["boundingRectangle"] as! [String: Any]
        self.boundingRectangle = InkBoundingRectangle(x: jsonBoundingRect["topX"] as! Float, y: jsonBoundingRect["topY"] as!Float,width: jsonBoundingRect["width"] as! Float,height: jsonBoundingRect["height"] as! Float)
        
        let jsonRotatedRectPoints = json["rotatedBoundingRectangle"] as! [[String: Any]]
        //Get the array of points for the rotated bounding rectangle and convert the millimeter values to CGFloats
        for point in jsonRotatedRectPoints {
            let x = point["x"] as! Float
            let y = point["y"] as! Float
            self.rotatedBoundingRectangle.append(CGPoint(x: InkPoint.millimeterToCGFloat(mmValue: x), y: InkPoint.millimeterToCGFloat(mmValue: y)))
        }
    }
}

File: java/InkRecognition/android-sample-app/Recognizer/app/src/main/java/CognitiveServices/Ink/Recognition/InkRecognitionUnit.java
package CognitiveServices.Ink.Recognition;

import android.util.DisplayMetrics;


import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class InkRecognitionUnit {

    private final ArrayList<Integer> strokeIds = new ArrayList<>();
    private int id;
    private InkRecognitionUnitCategory category;
    private final ArrayList<Integer> childIds = new ArrayList<>();
    private int parentId;
    private Rectangle boundingBox;
    private ArrayList<InkPoint> rotatedBoundingBox;
    private final float DOT_PER_INCH;
    private final InkRoot result;
    public static final float INCH_TO_MM = 25.4f;

    public InkRecognitionUnitCategory getCategory() {
        return category;
    }

    public ArrayList<Integer> getStrokeIds() {
        return strokeIds;
    }

    public int getId() {
        return id;
    }

    public ArrayList<InkRecognitionUnit> getChildren() {

        ArrayList<InkRecognitionUnit> childrenNodes;

        try {
            childrenNodes = result.getRecognitionUnits(childIds);
        }
        catch (Exception e) {
            System.out.print("Data Processing Error: " + e.getMessage());
            throw e;
        }
        return childrenNodes;
    }

    public InkRecognitionUnit getParent() {
        InkRecognitionUnit node;
        try {
            node = result.getRecognitionUnit(parentId);
        }
        catch (Exception e) {
            //An unknown data processing error occurred.
            System.out.print("Data Processing Error: " + e.getMessage());
            throw e;
        }
        return node;
    }

    public Rectangle getBoundingBox() {
        return boundingBox;
    }

    public ArrayList<InkPoint> getRotatedBoundingBox() {
        return rotatedBoundingBox;
    }

    InkRecognitionUnit(String jsonString, DisplayMetrics metrics, InkRoot result) throws JSONException {
        DOT_PER_INCH = metrics.xdpi;
        this.result = result;

        try {
            JSONObject obj = new JSONObject(jsonString);
            this.id = obj.getInt("id");
            this.category = getCategory(obj.getString("category"));
            this.parentId = obj.getInt("parentId");

            if (obj.has("childIds")) {
                JSONArray children = obj.getJSONArray("childIds");
                if (children.length() != 0) {
                    for (int i = 0; i < children.length(); i++) {
                        this.childIds.add(children.getInt(i));
                    }
                }
            }

            JSONArray strokes = obj.getJSONArray("strokeIds");
            if (strokes.length() != 0) {
                for (int i=0; i< strokes.length() ; i++) {
                    this.strokeIds.add(strokes.getInt(i));
                }
            }

            //setup bounding box
            JSONObject boundingRect = obj.getJSONObject("boundingRectangle");

            this.boundingBox = new Rectangle(millimetersToPixels((float)boundingRect.getDouble("topX")),
                    millimetersToPixels((float)boundingRect.getDouble("topY")),
                    millimetersToPixels((float)boundingRect.getDouble("width")),
                    millimetersToPixels((float)boundingRect.getDouble("height")));

            //setup rotated bounding box
            JSONArray rotatedBoundingRectPoints = obj.getJSONArray("rotatedBoundingRectangle");
            rotatedBoundingBox = new ArrayList<>();
            for (int i =0;
                 i <rotatedBoundingRectPoints.length();
                 i++) {
                JSONObject point = rotatedBoundingRectPoints.getJSONObject(i);
                this.rotatedBoundingBox.add(new InkPoint(millimetersToPixels((float)point.getDouble("x")), millimetersToPixels((float)point.getDouble("y"))));
            }
        }
        catch (Exception e) {
            System.out.print("Data Processing Error: " + e.getMessage());
            throw e;
        }
    }

    private InkRecognitionUnitCategory getCategory(String category) {

        InkRecognitionUnitCategory recognitionCategory;
        switch(category) {
            case "unknown":
                recognitionCategory = InkRecognitionUnitCategory.UNKNOWN;
                break;
            case "inkWord":
                recognitionCategory = InkRecognitionUnitCategory.INK_WORD;
                break;
            case "inkBullet":
                recognitionCategory = InkRecognitionUnitCategory.INK_BULLET;
                break;
            case "inkListItem":
                recognitionCategory = InkRecognitionUnitCategory.INK_LIST_ITEM;
                break;
            case "inkDrawing":
                recognitionCategory = InkRecognitionUnitCategory.INK_DRAWING;
                break;
            case "line":
                recognitionCategory = InkRecognitionUnitCategory.INK_LINE;
                break;
            case "paragraph":
                recognitionCategory = InkRecognitionUnitCategory.INK_PARAGRAPH;
                break;
            case "writingRegion":
                recognitionCategory = InkRecognitionUnitCategory.INK_WRITING_REGION;
                break;
                default:
                recognitionCategory = InkRecognitionUnitCategory.UNKNOWN;
                break;
        }
        return recognitionCategory;
    }

    float millimetersToPixels(float milliValue) {
        return (milliValue/INCH_TO_MM) * DOT_PER_INCH;
    }

}
